### PR TITLE
Add Ubuntu 24.04 noble support for OpenStack

### PR DESCRIPTION
**Pull request by danudey**
This patch adds preliminary support for building OpenStack packages to run on Ubuntu 24.04 Noble Numbat.

```release-note
Add Ubuntu 24.04 packages for OpenStack
```

Review notes:

1. Enable 'hide whitespace' when viewing the diff, as I did some whitespace consolidation (there were quite a few lines with tabs and spaces) but shell scripts aren't affected by whitespace changes so it doesn't need review.
2. 99% of the changes to the shell scripts themselves are just linting, and 99% of those changes involve putting quotes around variables and replacing sub-commands using backticks  with `$(command)` for readability, so you can safely ignore almost everything there

A few real changes to shell scripts:
1. The version parsing in `lib.sh` was a bit outdated so I fixed that up; it now creates valid versions for Python, and slightly-better versions for Ubuntu (though they still don't yet pass `lintian`'s standards).
2. I use dch to create/update the debian changelog, rather than generating it in the script directly; this feels more future-proof and less likely to cause issues down the road if something changes.

Packaging-related changes:
1. For Ubuntu packages, I made some changes to both the control files (specifically build dependencies) and to the shell script that installs ubuntu build dependencies, to remove old or outdated dependencies or update dependency versions, as some of it is no longer necessary in the versions of Ubuntu we support; for example, `dh-systemd` functionality is now part of `debhelper` in all current ubuntu versions, so we just update the version of `debhelper` we depend on and we can remove `dh-systemd`
2. We now depend on `debhelper-compat` so I removed `debian/compat` from felix and networking-calico
3. Likewise I removed the upstart scripts, since that hasn't been supported in ages
4. I ran `lintian` on packages and cleaned up what I could, though it does have some understandable complaints about how we do our packaging and our version numbers aren't quite up to snuff (they work, they just don't fit the proscribed pattern)
5. I added some file exclusions (to e.g. `felix/debian/source/options`) so that we're not packaging `Makefile`, the `.semaphore` directory, etc.

Dockerfile changes
1. I updated the dockerfiles so that they'll use a cache directory for all the files `apt` downloads, as well as bind-mounting the `install-ubuntu-build-deps` script in, preventing those files from being baked into the image. This will keep size down, which is just a nice to have for those cases where I forget to clean up my images
2. I fixed some deprecated syntax in the dockerfiles, solely to stop docker from complaining about it when we build the images

One thing I didn't change was removing Ubuntu Focal support; "standard support" for Focal ends this month (May 2025), though extended support continues into April 2030 for Ubuntu Pro subscribers; I doubt many of our customers are going to continue to use Focal for five more years and we might consider removing it soon, but for now there's no reason to take it out.

### CODE REVIEW

File: release/packaging/utils/create-update-packages.sh [hunk 1]
```diff
@@ -166,33 +166,33 @@ function precheck_pub_rpms {
 # Execution of the requested steps.
 
 function docker_run_rm {
-    docker run --rm --user $(id -u):$(id -g) -v $(dirname $(pwd)):/code -w /code/$(basename $(pwd)) "$@"
+    docker run --rm --user "$(id -u):$(id -g)" -v "$(dirname "$(pwd)"):/code" -w "/code/$(basename "$(pwd)")" "$@"
```
Comment (nelljerram) [inline release/packaging/utils/create-update-packages.sh:169]: Does the quoting really work in the 2nd and 3rd fragments here?

OK, it seems it does:
```
nell@glaurung:~/.banzai$ echo "$(id -u):$(id -g)" -v "$(dirname "$(pwd)"):/code" -w "/code/$(basename "$(pwd)")"     
1000:1000 -v /home/nell:/code -w /code/.banzai
```
I suppose that means `$( ... )` binds more strongly than double quotes.  I did not know that.

File: release/packaging/utils/create-update-packages.sh [hunk 2]
```diff
@@ -205,7 +205,6 @@ function do_felix {
       make build-bpf
       rm -f bpf-gpl/bin/test_*
     fi
-    rm -f Makefile
```
Comment (nelljerram) [inline release/packaging/utils/create-update-packages.sh:208]: Do you know why we did this?  I can't remember, but can probably do archaeology if needed.

Comment (nelljerram) [reply release/packaging/utils/create-update-packages.sh:208]: i.e. why we removed the Makefile and then reinstated it a few lines further down

File: release/packaging/utils/make-packages.sh [hunk 3]
```diff
@@ -39,114 +39,108 @@ for package_type in "$@"; do
...
+            patchelf --replace-needed libpcap.so.1 libpcap.so.0.8 bin/calico-felix
+        fi
+
+        # The Debian version that we are about to generate.
+        debver=${FORCE_VERSION_DEB:-$(git_version_to_deb "${version}")}
+        debver=$(strip_v "${debver}")
+
+        # Current time in Debian changelog format; e.g. Wed, 02
+        # Mar 2016 14:08:51 +0000.
+        timestamp=$(date "+%a, %d %b %Y %H:%M:%S %z")
```
Comment (nelljerram) [inline release/packaging/utils/make-packages.sh:56]: Still needed?

File: release/packaging/utils/make-packages.sh [hunk 4]
```diff
@@ -39,114 +39,108 @@ for package_type in "$@"; do
...
+        for series in focal jammy noble; do
+                if ${release}; then
+                        changelog_message="${NAME} v${debver} (from Git commit ${sha})."
+                else
+                        changelog_message="Development snapshot (from Git commit ${sha})."
+                fi
+                rm -f debian/changelog debian/changelog.dch
+                dch --controlmaint --create --package "${PKG_NAME}" -v "${DEB_EPOCH}${debver}-$series" "${changelog_message}"
+                dch --controlmaint --release ""
```
Comment (nelljerram) [inline release/packaging/utils/make-packages.sh:65]: Does this still put you as the author?  If so, I'm curious where that is configured.

**Review by nelljerram — APPROVED** [review]
Some comments and thoughts here.

Overall this looks like great stuff.  The proof of the pudding will, of course, be in whether the packages still build on Launchpad, and then whether the binary packages work and our OpenStack ST still passes - which of course we will keep an eye on.  Unfortunately those checks happen post-merge, so if there is any accidental breakage here we'll need to fix in a follow-on PR.  But I don't see anything here that looks like it _will_ be breakage.

File: release/packaging/docker-build-images/ubuntu-focal-build.Dockerfile.amd64 [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
-MAINTAINER Daniel Fox <dan.fox@tigera.io>
-ENV STREAM focal
+LABEL org.opencontainers.image.authors="Daniel Fox <dan.fox@tigera.io>"
+ENV STREAM=focal
 
-COPY install-ubuntu-build-deps install-ubuntu-build-deps
-RUN ./install-ubuntu-build-deps
+RUN \
+    --mount=type=cache,target=/var/cache/apt/archives \
+    --mount=type=cache,target=/var/lib/apt/lists \
```
[the remark this answers is not in the document]
Comment (danudey) [reply release/packaging/docker-build-images/ubuntu-focal-build.Dockerfile.amd64:7]: 1. It technically can, but I neglected to add in a `sharing` option - by default, all active builds will use the same cache, which, in the case of apt, does actually break. Adding `sharing=locked` makes them queue up, and `sharing=private` means everyone gets their own. Alternately, providing a unique cache ID will solve this if you *do* want it shared across *some* instances.
2. It helps me personally when I'm running things locally over and over again, and it doesn't cause any issues in production so I figured I may as well leave it in.

File: networking-calico/debian/control [hunk 6]
```diff
@@ -1,8 +1,8 @@
 Source: networking-calico
 Section: net
 Priority: optional
-Maintainer: Neil Jerram <neil@projectcalico.org>
```
[the remark this answers is not in the document]
Comment (danudey) [reply networking-calico/debian/control:4]: Throwback thursday!

File: felix/debian/source/options [hunk 7]
```diff
@@ -0,0 +1,2 @@
+tar-ignore = "Makefile"
+tar-ignore = ".semaphore"
```
[the remark this answers is not in the document]
Comment (danudey) [reply felix/debian/source/options:2]: 1. These paths are relative to "the files that dpkg is about to pack into the source tarball it's building", which, in effect, is relative to e.g. `<repo>/felix` in this case.
2. We actually already had code that would `rm Makefile`, run the build, and then `git checkout Makefile`, which excluded the `Makefile` from being packed in. I assume that this is in part because the `felix/` directory itself won't build without the rest of the source included and we don't necessarily want to ship the entire Calico repo as the source (though in the abstract that would make logical sense).

TL;DR the source package wasn't technically buildable originally, since it didn't include external-to-felix paths (e.g. it mounts `felix/../pod2daemon` into the go-build container, `../metadata.mk`, `../lib.Makefile`, etc)

My thoughts on this:

1. In an ideal world, we could build and install Calico without going out to any external sources; e.g. the build system that we have should be able to build in docker, podman, or just natively on the system using native libraries. If building natively, and you have the requirements installed, it wouldn't need to fetch docker images, etc.
2. If we had that, we could ship an actual source package and build an actual binary package from it, rather than the fake-out that we do now.
3. In that ideal world, we could build Calico components like felix, calicoctl, etc., on a given system and use that system's native libraries if we (or our users) wanted to (not that golang apps use native libraries outside of `CGO`, etc.)
5. This is likely never going to happen

Part of that would be just the sheer amount of work it would take to untangle everything and remove all of our assumptions, of course; the other part is that it makes very little sense, since 99% of the time what we're actually trying to build is docker containers, and it makes no sense to go to all this trouble to make "native builds" for not-actually-native build artifacts.

My biggest concern is that, when we ship our "source" packages for Calico to launchpad, we're not actually shipping all the source; we're only shipping the part of it that exists in the `felix/` directory. I don't think we have any *legal* exposure here, but I do think:
1. This kind of goes against the whole point of using Launchpad to distribute "open source" packages; even if that code is *technically* available, we're not uploading it, and a user downloading the 'source' package isn't going to get what they think they're going to get.
2. There's not really a sensible way of resolving that without a lot of work

I mean, theoretically we could just run our own package repositories from Google Artifact Registry or PackageCloud or something, and then stop pretending that our source packages are source packages. That would be fine too.

File: release/packaging/utils/create-update-packages.sh [hunk 8]
```diff
@@ -107,7 +107,7 @@ function require_deb_secret_key {
 # Decide target arch; by default the same as the native arch here.  We
 # conventionally say "amd64", where uname says "x86_64".
 ARCH=${ARCH:-$(uname -m)}
-if [ $ARCH = x86_64 ]; then
+if [ "$ARCH" = x86_64 ]; then
```
[the remark this answers is not in the document]
Comment (danudey) [reply release/packaging/utils/create-update-packages.sh:110]: Less about being non-empty and more about how weirdly things would break if there was a space in the output. Though that, in and of itself, would be extremely weird.

File: release/packaging/utils/create-update-packages.sh (hunk 1 above)
Comment (danudey) [reply release/packaging/utils/create-update-packages.sh:169]: Yeah, this is one of the reasons I switch a lot of things to `$(...)`. Also it makes it far, far easier to nest things (like in this case).

File: release/packaging/utils/create-update-packages.sh (hunk 2 above)
Comment (danudey) [reply release/packaging/utils/create-update-packages.sh:208]: I assume (see above discussion) so as not to include it in the source package, since it won't work without the rest of the source from outside of the `felix/` directory.

File: release/packaging/utils/make-packages.sh (hunk 3 above)
Comment (danudey) [reply release/packaging/utils/make-packages.sh:56]: Not at all!

File: felix/debian/source/options (hunk 7 above)
[the remark this answers is not in the document]
Comment (nelljerram) [reply felix/debian/source/options:2]: Thanks, we have the same understanding about all this.  I can't remember precisely why we started pre-building the felix executable.  It might have been just laziness, but I think it was probably because some required build-dep wasn't available in Ubuntu when we first encoded this.  If so, that might have been fixed since... but as you say, it would still be a lot of someone's time.

Regarding the Makefile dance, I think it was something like: if a `Makefile` is present, `dpkg-buildpackage` can infer things from certain targets within it; but in our case those inferences were wrong and caused problems, so I worked around that by deleting the Makefile before `dpkg-buildpackage` and reinstating it afterwards.  Anyway, we can see if things now work without this dance; hopefully so.

File: release/packaging/docker-build-images/ubuntu-focal-build.Dockerfile.amd64 (hunk 5 above)
[the remark this answers is not in the document]
Comment (nelljerram) [reply release/packaging/docker-build-images/ubuntu-focal-build.Dockerfile.amd64:7]: OK, (2) is actually a very strong point!  So definitely worth keeping this.  I don't entirely follow the `sharing` point, but presume that you've got it right now.

File: release/packaging/utils/create-update-packages.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (nelljerram) [reply release/packaging/utils/create-update-packages.sh:110]: Right, excellent point!

File: felix/debian/source/options (hunk 7 above)
[the remark this answers is not in the document]
Comment (danudey) [reply felix/debian/source/options:2]: > if a Makefile is present, dpkg-buildpackage can infer things from certain targets within it

Aah, this makes sense, yep. Also, Ubuntu's build system is philosophically against building stuff in containers (that defeats the whole purpose) so I would guess that running builds in Docker on Launchpad is probably not practical/possible either.
